examples/python/learning_pytorch.py

```diff
@@ -48,230 +47,250 @@
 # config_file_path = "../../scenarios/rocket_basic.cfg"
 # config_file_path = "../../scenarios/basic.cfg"
 
-# Converts and down-samples the input image
+# Uses GPU if available
+if torch.cuda.is_available():
+    DEVICE = torch.device('cuda')
+    torch.backends.cudnn.benchmark = True
+else:
+    DEVICE = torch.device('cpu')
+
+
 def preprocess(img):
+    """Down samples image to resolution"""
     img = skimage.transform.resize(img, resolution)
     img = img.astype(np.float32)
+    img = np.expand_dims(img, axis=0)
     return img
 
 
-class ReplayMemory:
-    def __init__(self, capacity):
-        channels = 1
-        state_shape = (capacity, channels, resolution[0], resolution[1])
-        self.s1 = np.zeros(state_shape, dtype=np.float32)
-        self.s2 = np.zeros(state_shape, dtype=np.float32)
-        self.a = np.zeros(capacity, dtype=np.int32)
-        self.r = np.zeros(capacity, dtype=np.float32)
-        self.isterminal = np.zeros(capacity, dtype=np.float32)
+def create_simple_game():
+    print("Initializing doom...")
+    game = vzd.DoomGame()
+    game.load_config(config_file_path)
+    game.set_window_visible(False)
+    game.set_mode(Mode.PLAYER)
+    game.set_screen_format(vzd.ScreenFormat.GRAY8)
+    game.set_screen_resolution(vzd.ScreenResolution.RES_640X480)
+    game.init()
+    print("Doom initialized.")
+
+    return game
+
+
+def test(game, agent):
+    """Runs a test_episodes_per_epoch episodes and prints the result"""
+    print("\nTesting...")
+    test_scores = []
+    for test_episode in trange(test_episodes_per_epoch, leave=False):
+        game.new_episode()
+        while not game.is_episode_finished():
+            state = preprocess(game.get_state().screen_buffer)
+            best_action_index = agent.get_action(state)
+
+            game.make_action(actions[best_action_index], frame_repeat)
+        r = game.get_total_reward()
+        test_scores.append(r)
+
+    test_scores = np.array(test_scores)
+    print("Results: mean: %.1f +/- %.1f," % (
+        test_scores.mean(), test_scores.std()), "min: %.1f" % test_scores.min(),
+          "max: %.1f" % test_scores.max())
+
+
+def run(game, agent, actions, num_epochs, frame_repeat, steps_per_epoch=2000):
+    """
+    Run num epochs of training episodes.
+    Skip frame_repeat number of frames after each action.
+    """
+    scores = []
+    start_time = time()
+
+    for epoch in range(num_epochs):
+        game.new_episode()
+        train_scores = []
+        global_step = 0
+        print("\nEpisode #" + str(epoch + 1))
```

Comment:
Replace with "Epoch" (one episode is one play through the game from spawning to end, so this is confusing).

Reply:
Done